### PR TITLE
SFD-2789 terraform iam roles for service accounts

### PR DESCRIPTION
**Pull request by gwaardenburg**
You can ignore the changes in the cognito.sh file. I had to replace the tabs with double spaces to get passed the commit verification messages.

### CODE REVIEW

File: deployment/stacks/service/base/template/service-search/service-account.yaml [hunk 1]
```diff
@@ -3,4 +3,4 @@ kind: ServiceAccount
 metadata:
   name: uec-dos-api-sfs-service-account
   annotations:
-    eks.amazonaws.com/role-arn: arn:aws:iam::AWS_ACCOUNT_ID_TO_REPLACE:role/uec-dos-api-sfs-pod
+    eks.amazonaws.com/role-arn: arn:aws:iam::AWS_ACCOUNT_ID_TO_REPLACE:role/SERVICE_TAG_TO_REPLACE-PROFILE_TO_REPLACE-PROJECT_TAG_TO_REPLACE-dynamodb-access
```
Comment (jope1) [inline deployment/stacks/service/base/template/service-search/service-account.yaml:6]: SFS may need more than just access to Dynamo in future.  Good idea having the PROFILE in the name though. Can we please use:
SERVICE_TAG_TO_REPLACE-PROFILE_TO_REPLACE-SA

File: infrastructure/stacks/roles/locals.tf [hunk 2]
```diff
@@ -0,0 +1,15 @@
+locals {
+
```
Comment (jope1) [inline infrastructure/stacks/roles/locals.tf:2]: Remove extra blank line please

File: infrastructure/stacks/roles/locals.tf [hunk 3]
```diff
@@ -0,0 +1,15 @@
...
+
+  fuzzy-search-dynamodb_access_iam_name = "${var.service_tag}-${var.profile}-${var.project_tag}-dynamodb-access"
```
Comment (jope1) [inline infrastructure/stacks/roles/locals.tf:4]: Needs to align with the K8s service account name.  And please call this service-account-role or service-account-iam-role

File: infrastructure/stacks/roles/locals.tf [hunk 4]
```diff
@@ -0,0 +1,15 @@
...
+
+  dynamoDb_full_access_policy_arn      = "arn:aws:iam::aws:policy/AmazonDynamoDBFullAccess"
```
Comment (jope1) [inline infrastructure/stacks/roles/locals.tf:6]: Fuzzy API only needs read-only access. Is there an option for this?  Also, there should be a way of only allowing the API to reach our DynamoDB

File: infrastructure/stacks/roles/main.tf [hunk 5]
```diff
@@ -0,0 +1,24 @@
+
+resource "aws_iam_role" "fuzzy-search-dynamodb_access" {
```
Comment (jope1) [inline infrastructure/stacks/roles/main.tf:2]: Please bring alias name inline with above

File: infrastructure/stacks/roles/variables.tf [hunk 6]
```diff
@@ -0,0 +1,32 @@
...
+
+variable "security_groups_terraform_state_key" { description = "The security groups key in the terraform state bucket" }
+
+variable "security_groups_k8s_terraform_state_key" { description = "The k8s security groups key in the terraform state bucket" }
+
+variable "programme" { description = "The programme tag" }
+
+variable "service_tag" { description = "The service tag" }
+
+variable "project_tag" { description = "The project tag" }
```
Comment (jope1) [inline infrastructure/stacks/roles/variables.tf:32]: Do we need all of these variables?

**Review by jope1 — CHANGES_REQUESTED** [review]

File: infrastructure/stacks/roles/variables.tf (hunk 6 above)
Comment (gwaardenburg) [reply infrastructure/stacks/roles/variables.tf:34]: I think so as they are pointing to the general terraform infrastructure bucket. Unless I am mistaken?

File: infrastructure/stacks/roles/locals.tf (hunk 4 above)
Comment (gwaardenburg) [reply infrastructure/stacks/roles/locals.tf]: So I added the read only dynamoDB out the box policy, however I then couldnt lock it down to a specific table. So what I have done is I have copied the read only policy and created a new one, but locked that new one down to our specific table.

**Review by jope1 — APPROVED** [review]

**Review by ArifNhs — APPROVED** [review]
Approved

**Pull request merged**